include/libhal-soft/bit_bang_i2c.hpp

```diff
@@ -0,0 +1,177 @@
+// Copyright 2024 Khalil Estell
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <chrono>
+#include <libhal/i2c.hpp>
+#include <libhal/output_pin.hpp>
+#include <libhal/units.hpp>
+
+namespace hal {
+/**
+ * @brief A bit bang implementation for i2c.
+ *
+ * This implementation of i2c only needs 2 gpios and a steady_clock to work
+ * correctly. It supports single-controller multiple peripherals but  we intend
+ * to support multiple controllers in the future. bit-bang is a best-effort
+ * implementation of i2c. The maximum achievable clock rate for the lpc4078 is
+ * about 180kHz. Interrupts disrupt this controller.
```

Comment:
Explain why interrupts disrupt the controller for developers who may not fully understand the implications.